### DISCUSSION TITLE
Why bcrypt.compare function no need to use salt, when compare with origin password?

### DISCUSSION THREAD

**Discussion by xgqfrms** [kelektiv/node.bcrypt.js#972] [Q&A]
# Why bcrypt.compare function no need to use salt, when compare with origin password?

`hash-salt-test.js`

```js
const bcrypt = require('bcrypt');

const hash = async function (password, salt = 10) {
  // ❓default salt: 10
  console.log(`❓salt =`, salt);
  const hashed = await bcrypt.hash(password, salt);
  return hashed;
}

const check = async function (password, hashed) {
  // ❓why no need to use any salt when call compare function
  const result = await bcrypt.compare(password, hashed);
  return result;
}

const test = async function (salt = 10) {
  // password length is 10 ✅, less than 72 bytes ❓
  const pwd = 'pwd1234567';
  const hashed = await hash(pwd, salt);
  const result = await check(pwd, hashed);
  console.log(`\nhash with salt`, hashed);
  console.log(`result =`, result);
}

test();
test(5);
test(11);
test(15);
test(17);

/*

 $ node ./hash-salt-test.js

❓salt = 10
❓salt = 5
❓salt = 11
❓salt = 15
❓salt = 17

hash with salt $2b$05$.zICWxt0wAjGb58kIN6GwekMtP5jUeYD34/wqfE17yYqZ3rO43RV.
result = true

hash with salt $2b$10$PPJdjtaboGieRgk1p86bsuCHIqmiFhqaRGwHrpVs5/dZ3ZLC8b33m
result = true

hash with salt $2b$11$Y.Ozy9S5gtZUR1XD3egrGeZ1H8HmPOUqBDochpGM5pLCIP2tk38Xa
result = true

hash with salt $2b$15$UBGFkZRAxGgMMTaRDmxJ1eTgCBM.ovvESEyJJwrGeIpk/Vw3NPFkC
result = true

hash with salt $2b$17$zyjemy4KDsQDO8Q2kCgYeeW/FsG3kiWp0hXl4t.o8GUWrWjA4.R.S
result = true

*/
```

**Comment by xgqfrms** [comment]
![image](https://user-images.githubusercontent.com/7291672/206617849-489dbfb5-a45a-498a-9808-ed4aa01ccde9.png)

https://github.com/kelektiv/node.bcrypt.js/issues/963#issuecomment-1272338945

**Comment by xgqfrms** [comment]
Per bcrypt implementation, only `the first 72 bytes` of a string are used. 
Any extra bytes are `ignored` when matching passwords. 

Note that this is not the first 72 characters. 
It is possible for a string to `contain less than` 72 characters, while `taking up more than` 72 bytes 
(e.g. a `UTF-8` encoded string containing `emojis`).

https://github.com/kelektiv/node.bcrypt.js/issues/935#issuecomment-1159411386

**Comment by recrsn** [comment]
Bcrypt generates the hashed password in modular crypt format (MCF) the hash itself contains information about the algorithm, rounds and the salt
